File: CPP/countingSort.cpp
// Counting Sort for non-negative numbers

#include <iostream>
#include <vector>
// vector is required for creating buckets.
// because the size of the array is not known at compile time.

using namespace std;

template<size_t N>
int findMax(int (&arr)[N]);

template<size_t N>
void countingSort(int (&arr)[N]);


int main() {
    int myArray[] = {4, 2, 2, 6, 3, 3, 1, 6, 5, 2, 3};
    countingSort(myArray);

    for (int num : myArray) {
        cout << num << " ";
    }
    cout << endl;
    
    return 0;
}

template<size_t N>
void countingSort(int (&arr)[N]) {
    // Create bucket for each number (including 0)
    const int length = findMax(arr) + 1;
    vector<int> buckets(length, 0);

    // Count the number of each element in the array
    for (int i : arr) {
        buckets[i]++;
    }

    // Sort the array
    int index = 0;
    for (int bucket = 0; bucket < length; ++bucket) {
        for (int n = 0; n < buckets[bucket]; ++n) {
            // `index++` increments the index after the value is assigned
            arr[index++] = bucket;
        }
    }

}

template<size_t N>
int findMax(int (&arr)[N]) {
    // Returns maximum number of the given array
    int maxNum = arr[0];

    for (int i : arr) {
        if (i > maxNum) {
            maxNum = i;
        }
    }

    return maxNum;
}

File: Python/counting_sort.py
'''
Counting Sort

Sorts by counting the number of time each value occurs.  
Fast when the range of possible values < number of values.  
Otherwise, the array created for counting will be larger than the array being sorted.  
Works for non-negative integers.

Tested with Python 3.13.2
'''

def counting_sort(arr: list):
    max_num = max(arr) # How many numbers are required
    buckets = [0] * (max_num+1) # Create a bucket for each number

    # Count the number of times each number appears
    for i in arr:
        buckets[i] += 1

    arr.clear() # Clear the original array

    # Add the numbers back to the array in order
    for i, num in enumerate(buckets):
        for _ in range(num):
            arr.append(i)

# USAGE
my_array = [4, 2, 2, 6, 3, 3, 1, 6, 5, 2, 3]
counting_sort(my_array)
print(my_array)